settings: its: enable tf-m PSA headers and run more tests

dsseng opened the pull request:
- **settings: its: enable tf-m PSA headers**
- **tests: settings: its: run on mps2/an521/cpu0/ns**

dsseng committed:
settings: its: enable TF-M PSA headers

Avoid missing headers on some platforms like mps2/an521/cpu0/ns.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: settings: its: run on mps2/an521/cpu0/ns

Run on QEMU to make sure this is tested in the CI.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented on subsys/settings/src/CMakeLists.txt:
```diff
@@ -16,3 +16,8 @@ zephyr_sources_ifdef(CONFIG_SETTINGS_SHELL settings_shell.c)
 zephyr_sources_ifdef(CONFIG_SETTINGS_TFM_ITS settings_its.c)
 zephyr_sources_ifdef(CONFIG_SETTINGS_RETENTION settings_retention.c)
+
+if (CONFIG_SETTINGS_TFM_ITS)
+  # For psa/internal_trusted_storage.h in case of TF-M implementation
+  zephyr_library_link_libraries_ifdef(CONFIG_BUILD_WITH_TFM tfm_api)
+endif()
```
mbedtls and entropy modules do the same. There's also a more verbose option which I preferred not to use since CMake can handle it all with link_libraries:

```cmake
  if (CONFIG_BUILD_WITH_TFM)
    zephyr_library_include_directories(
      $<TARGET_PROPERTY:tfm,TFM_BINARY_DIR>/api_ns/interface/include
    )
  endif()
```

dsseng committed:
samples: settings: run integration on QEMU with TF-M

Make sure ITS backend is tested in the CI.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: settings: its: enable functional tests for TF-M ITS

Run on mps2/an521/cpu0/ns.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented on samples/subsys/settings/sample.yaml:
```diff
@@ -19,31 +19,46 @@ tests:
+common:
+  timeout: 20
+  harness: console
+  harness_config:
```
Thanks, re-checked, it times out when I add some extra line, so looks good

tomi-font commented on tests/subsys/settings/functional/psa/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_ZTEST=y
+
+CONFIG_SETTINGS=y
+CONFIG_SETTINGS_RUNTIME=y
```
Why this?

dsseng replied:
dropped

tomi-font commented on tests/subsys/settings/functional/psa/settings_test_psa.c:
```diff
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright (c) 2022 Dmitrii Sharshakov <d3dx12.xx@gmail.com> */
```
2025

tomi-font commented on tests/subsys/settings/functional/psa/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+tests:
+  settings.functional.psa-its:
+    platform_allow:
+      - mps2/an521/cpu0/ns
```
Same comments than for the sample.

dsseng replied:
Please see here: https://github.com/zephyrproject-rtos/zephyr/actions/runs/17334214787/job/49217353543?pr=94881#step:13:3052

Either we should get max size from TF-M config, or only enable building the test for those platforms we explicitly configure.

I think we should use TF-M config as the source, because it's weird to use the TF-M config override option here

tomi-font commented on tests/subsys/settings/functional/src/settings_basic_test.c:
```diff
@@ -38,7 +43,16 @@ LOG_MODULE_REGISTER(settings_basic_test);
+	psa_status_t status;
+
+	/* Remove all potentially accessed ITS entries in the UID range */
+	for (int i = 0; i < (sizeof(entries) / CONFIG_TFM_ITS_MAX_ASSET_SIZE) + 1; i++) {
```
extraneous parentheses

tomi-font commented on tests/subsys/settings/functional/src/settings_basic_test.c:
```diff
@@ -38,7 +43,16 @@ LOG_MODULE_REGISTER(settings_basic_test);
+	/* Remove all potentially accessed ITS entries in the UID range */
+	for (int i = 0; i < (sizeof(entries) / CONFIG_TFM_ITS_MAX_ASSET_SIZE) + 1; i++) {
+		status = psa_its_remove(ZEPHYR_PSA_SETTINGS_TFM_ITS_UID_RANGE_BEGIN + i);
+		zassert_true((status == PSA_SUCCESS) || (status == PSA_ERROR_DOES_NOT_EXIST),
```
extraneous parentheses

tomi-font commented on subsys/settings/src/settings_its_priv.h:
```diff
@@ -0,0 +1,17 @@
+	size_t val_len;
+};
+
+static struct setting_entry entries[CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES];
```
Having `static struct setting_entry entries` in the header file is wrong as we'll end up with multiple definitions of `entries` for every source file including this header. It's not that big of a problem right now as the test which includes this just uses `sizeof(entries)`, but it's still not great.
Actually to make matters simpler why not just remove this line and move it back to `settings_its.c` and in the test use `sizeof(struct settings_entry) * CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES`?

dsseng replied:
should work, thanks

tomi-font reviewed (changes requested):
Nice work! Some not-so-major comments.

dsseng commented on subsys/settings/src/CMakeLists.txt:
```diff
@@ -16,3 +16,8 @@ zephyr_sources_ifdef(CONFIG_SETTINGS_SHELL settings_shell.c)
 zephyr_sources_ifdef(CONFIG_SETTINGS_RETENTION settings_retention.c)
+
+if (CONFIG_SETTINGS_TFM_ITS)
+  # For psa/internal_trusted_storage.h in case of TF-M implementation
```
considered this redundant, yes, thanks

dsseng commented on samples/subsys/settings/sample.yaml:
```diff
@@ -19,31 +49,14 @@ tests:
+      - mps2/an521/cpu0/ns
+    extra_args:
+      - CONFIG_SETTINGS_TFM_ITS=y
+      - CONFIG_TFM_ITS_MAX_ASSET_SIZE_OVERRIDE=y
```
Setting it in a board-specific config file, since this might be different. Perhaps 512 is a safe default (less than TF-M gives us is OK), but let's keep it in board configs maybe?

I think we could use the option from TF-M config here actually: https://github.com/zephyrproject-rtos/zephyr/pull/94947/files#diff-127df73a3b182f206edf295cb6dd7520f2c85394a1e7d0d9f90963ab5454f89cR37. What do you think?

dsseng replied:
This should work, I haven't seen less than 512 in TF-M upstream. It's still a bit weird to me that we require an *override* for basic functionality. IIRC the override option is to change the TF-M config, which most users should not touch without a need. Why cannot we use the setting value TF-M itself exposes?

dsseng replied:
This is what I came up with, should cover both tfm and secure storage options

```c
#if defined(CONFIG_BUILD_WITH_TFM)
/* For ITS_MAX_ASSET_SIZE */
#include "config_base.h"
#else
#define ITS_MAX_ASSET_SIZE CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE
#endif
```

tomi-font replied:
Just keep it to `#include "config_base.h"`, this Settings backend is only for TF-M:
https://github.com/zephyrproject-rtos/zephyr/blob/3a54a4d1239fbe795af978d4c75cd91b87fd73cd/subsys/settings/Kconfig#L127-L129
Making it usable with the Secure storage subsystem is a different topic that would need a discussion of its own. It's not the original intent.

dsseng committed:
settings: its: move structure to a header

This private header can be used by tests for information on UID ranges
that should be cleaned for testing.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented on tests/subsys/settings/functional/psa/CMakeLists.txt:
See #94947 - there is a PS backend based on mostly the same sources and principles. I can rename it here and rename in the later PR though if desired

tomi-font replied:
Right, no need to rename things back and forth but I would still like to have TF-M in the name to make things clear, so maybe `tfm_psa`?

dsseng commented on samples/subsys/settings/sample.yaml:
```diff
@@ -1,10 +1,40 @@
+        \ key <alpha/beta/source>"
+
 tests:
   sample.subsys.settings:
```
I guess this change would be better reverted altogether, as samples' tests are for testing samples, not the subsystems. Thus, sample test must not cover all the cases subsys abstracts out. At least this is how I got that

tomi-font replied:
Hmm yeah, as you want, both are fine to me. It's true we should at least keep a working default configuration, but just adding more configurations for CI to me is totally fine.

tomi-font commented on tests/subsys/settings/functional/psa/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+    tags:
+      - settings
+      - trusted-firmware-m
+    extra_args:
+      - CONFIG_SETTINGS_TFM_ITS=y
+      - CONFIG_TFM_ITS_MAX_ASSET_SIZE_OVERRIDE=y
```
Alright. Though `CONFIG_TFM_ITS_MAX_ASSET_SIZE_OVERRIDE` could still be moved as I see in https://github.com/zephyrproject-rtos/zephyr/pull/94947 you add it to both tests.

dsseng committed:
settings: its: do not rely on TFM settings override

Use config from TF-M instead, not necessitating an override.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented:
sorry, I'm surprized with myself taking 2w to fix the wording...

tomi-font commented on tests/subsys/settings/functional/src/settings_basic_test.c:
```diff
@@ -24,6 +23,17 @@ LOG_MODULE_REGISTER(settings_basic_test);
+#include "config_base.h"
+#else
+#define ITS_MAX_ASSET_SIZE CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE
+#endif
```
No, the Settings backend is only for TF-M's ITS. It doesn't make sense to enable it for Secure Storage. Plus with it would come potential infinite call recursion (Settings => Secure Storage => Settings...).
https://github.com/zephyrproject-rtos/zephyr/pull/94881#discussion_r2368046752

tomi-font commented on subsys/settings/src/CMakeLists.txt:
```diff
@@ -16,3 +16,7 @@ zephyr_sources_ifdef(CONFIG_SETTINGS_SHELL settings_shell.c)
 zephyr_sources_ifdef(CONFIG_SETTINGS_RETENTION settings_retention.c)
+
+if(CONFIG_SETTINGS_TFM_ITS)
+  zephyr_library_link_libraries_ifdef(CONFIG_BUILD_WITH_TFM tfm_api)
```
Hmm kinda same thing here, the `CONFIG_BUILD_WITH_TFM` condition will always be true within this `if` so you can just link the `tfm_api` library unconditionally.

dsseng commented on subsys/settings/src/settings_its.c:
```diff
@@ -12,17 +12,20 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/psa/its_ids.h>
 
+#include "settings_its_priv.h"
```
I don't think so, here it is a local file in the same directory, so let's start searching for it in the same directory

dsseng commented on tests/subsys/settings/its/prj.conf:
Will follow up with this in the PS PR anyway
